tree/treeplayer/inc/TTreeReader.h

```diff
@@ -78,10 +78,28 @@ class TTreeReader: public TObject {
          fEntry(entry), fReader(&reader) {}
 
       /// Compare two iterators for equality.
-      bool operator==(const Iterator_t& lhs) const {
+      bool operator==(const Iterator_t &lhs) const {
          // From C++14: value initialized (past-end) it compare equal.
-         if (!IsValid() && !lhs.IsValid()) return true;
-         return fEntry == lhs.fEntry && fReader == lhs.fReader;
+         if (!IsValid() && !lhs.IsValid())
+            return true;
+         // The iterators refer to different readers
+         if (fReader != lhs.fReader)
+            return false;
+         // #16249: range based loop and the tree has zero entries
+         // as well as analogous cases.
+         // Getting the number of events can have a cost, for example in
+         // case of chains of remote files accessible with high latency.
+         // However, it is reasonable to assume that if iterators are
+         // being compared is because an iteration is taking place,
+         // therefore such cost has to be paid anyway, it's just
+         // anticipated.
+         if (fReader->GetTree()->GetEntries() == 0 && fEntry == 0 && !lhs.IsValid()) {
+            return true;
+         }
+         if (lhs.fReader->GetTree()->GetEntries() == 0 && lhs.fEntry == 0 && !IsValid()) {
+            return true;
+         }
```

Comment:
```suggestion
         if (fReader->GetTree()->GetEntriesFast() == 0 && fEntry == 0 && !lhs.IsValid()) {
            return true;
         }
         if (lhs.fReader->GetTree()->GetEntriesFast() == 0 && lhs.fEntry == 0 && !IsValid()) {
            return true;
         }
```
If `LoadTree(0)` has been called and there is really zero elements in the chain then `GetEntriesFast` will be zero.
If there is code path where we can get here and `LoadTree(0)` has not been called I suspect we could call it safely here (but hopefully we don't have this case).